File: 프로그래머스/lv1/67256. ［카카오 인턴］ 키패드 누르기/［카카오 인턴］ 키패드 누르기.py
def compare_distance(n, l, r, hand):
    l_distance = abs(l[0] - n[0]) + abs(l[1] - n[1])
    r_distance = abs(r[0] - n[0]) + abs(r[1] - n[1])
    if l_distance > r_distance:
        return 'R'
    elif l_distance < r_distance:
        return 'L'
    else:
        if hand == 'left':
            return 'L'
        else: return 'R'

def solution(numbers, hand):
    result = []
    keypad = {1:(0,0), 2:(0,1), 3:(0,2), 4:(1,0), 5:(1,1), 6:(1,2), 7:(2,0), 8:(2,1), 9:(2,2), '*':(3,0), 0:(3,1), '#':(3,2),}
    left_cur = '*'
    right_cur = '#'
    
    for num in numbers:
        if num in [1, 4, 7]:
            result.append('L')
            left_cur = num
        elif num in [3, 6, 9]:
            result.append('R')
            right_cur = num
        else: # 2, 5, 8, 0
            l = keypad[left_cur]
            r = keypad[right_cur]
            n = keypad[num]
            if compare_distance(n, l, r, hand) == 'L':
                result.append('L')
                left_cur = num
            else:
                result.append('R')
                right_cur = num
                
    return ''.join(result)
            
            
            
            
            
            
                
        

File: 프로그래머스/1/67256. ［카카오 인턴］ 키패드 누르기/［카카오 인턴］ 키패드 누르기.java
import java.util.HashMap;

class Solution {
    static HashMap<Integer, int[]> row;
    public String solution(int[] numbers, String hand) {
        String answer = "";
        row = new HashMap<>();
        for (int i = 0; i <= 2; i++) {
            for (int j = (i * 3) + 1; j <= (i * 3) + 3; j++) {
                if (j % 3 == 0) row.put(j, new int[]{i+1, (j % 3) + 3});
                else row.put(j, new int[]{i+1, j % 3});
            }
        }
        row.put(0, new int[]{4, 2}); row.put(10, new int[]{4, 1});
        row.put(20, new int[]{4, 3});

        int leftThumb = 10, rightThumb = 20;

        for (int num : numbers) {
            if (num % 3 == 1) {
                answer += "L";
                leftThumb = num;
            } else if (num != 0 && (num % 3 == 0)) {
                answer += "R";
                rightThumb = num;
            } else {
                int leftDistance = Math.abs(row.get(leftThumb)[0] - row.get(num)[0]) + Math.abs(row.get(leftThumb)[1] - row.get(num)[1]);
                int rightDistance = Math.abs(row.get(rightThumb)[0] - row.get(num)[0]) + Math.abs(row.get(rightThumb)[1] - row.get(num)[1]);
                if (leftDistance < rightDistance) {
                    answer += "L";
                    leftThumb = num;
                } else if (rightDistance < leftDistance) {
                    answer += "R";
                    rightThumb = num;
                } else {
                    if (hand.equals("left")) {
                        answer += "L";
                        leftThumb = num;
                    } else {
                        answer += "R";
                        rightThumb = num;
                    }
                }
            }
        }

        return answer;
    }
}